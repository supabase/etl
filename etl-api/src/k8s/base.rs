use async_trait::async_trait;
use etl_config::Environment;
use k8s_openapi::api::core::v1::ConfigMap;
use thiserror::Error;

/// Errors emitted by the Kubernetes integration.
///
/// Variants wrap lower-level libraries where appropriate to preserve context.
#[derive(Debug, Error)]
pub enum K8sError {
    /// A serialization or deserialization error while building or parsing
    /// Kubernetes resources.
    #[error("An error occurred in serde when dealing with K8s: {0}")]
    Serde(#[from] serde_json::error::Error),
    /// An error returned by the [`kube`] client when talking to the API
    /// server.
    #[error("An error occurred with kube when dealing with K8s: {0}")]
    Kube(#[from] kube::Error),
}

/// A simplified view of a pod phase.
///
/// This mirrors the string phases reported by Kubernetes but only tracks the
/// states needed by the API. Unknown values map to [`PodPhase::Unknown`].
pub enum PodPhase {
    Pending,
    Running,
    Succeeded,
    Failed,
    Unknown,
}

impl From<&str> for PodPhase {
    /// Converts a Kubernetes pod phase string into a [`PodPhase`].
    ///
    /// Unrecognized values result in [`PodPhase::Unknown`].
    fn from(value: &str) -> Self {
        match value {
            "Pending" => PodPhase::Pending,
            "Running" => PodPhase::Running,
            "Succeeded" => PodPhase::Succeeded,
            "Failed" => PodPhase::Failed,
            _ => PodPhase::Unknown,
        }
    }
}

/// Client interface describing the Kubernetes operations used by the API.
///
/// Implementations are expected to be idempotent where possible by issuing
/// server-side apply patches for create-or-update behaviors.
#[async_trait]
pub trait K8sClient: Send + Sync {
    /// Creates or updates the Postgres password secret for a replicator.
    ///
    /// The secret name is derived from `prefix` and is stored in the
    /// data-plane namespace.
    async fn create_or_update_postgres_secret(
        &self,
        prefix: &str,
        postgres_password: &str,
    ) -> Result<(), K8sError>;

    /// Creates or updates the BigQuery service account secret for a
    /// replicator.
    ///
    /// The secret name is derived from `prefix` and is stored in the
    /// data-plane namespace.
    async fn create_or_update_bq_secret(
        &self,
        prefix: &str,
        bq_service_account_key: &str,
    ) -> Result<(), K8sError>;

    /// Deletes the Postgres password secret for a replicator if it exists.
    async fn delete_postgres_secret(&self, prefix: &str) -> Result<(), K8sError>;

    /// Deletes the BigQuery service account secret for a replicator if it
    /// exists.
    async fn delete_bq_secret(&self, prefix: &str) -> Result<(), K8sError>;

    /// Retrieves a named [`ConfigMap`].
    async fn get_config_map(&self, config_map_name: &str) -> Result<ConfigMap, K8sError>;

    /// Creates or updates the replicator configuration [`ConfigMap`].
    ///
    /// The config map stores two YAML documents: a base and a environment specific
    /// override.
    async fn create_or_update_config_map(
        &self,
        prefix: &str,
        base_config: &str,
        env_config: &str,
        environment: Environment,
    ) -> Result<(), K8sError>;

    /// Deletes the replicator configuration [`ConfigMap`] if it exists.
    async fn delete_config_map(&self, prefix: &str) -> Result<(), K8sError>;

    /// Creates or updates the replicator [`StatefulSet`].
    ///
    /// The set references previously created secrets and config maps. Optional
    /// `template_annotations` may be used to trigger a rolling restart.
    async fn create_or_update_stateful_set(
        &self,
        prefix: &str,
        replicator_image: &str,
        environment: Environment,
    ) -> Result<(), K8sError>;

    /// Deletes the replicator [`StatefulSet`] if it exists.
    async fn delete_stateful_set(&self, prefix: &str) -> Result<(), K8sError>;

    /// Returns the phase of the replicator pod.
    async fn get_pod_phase(&self, prefix: &str) -> Result<PodPhase, K8sError>;

    /// Reports whether the replicator container terminated with a non-zero exit
    /// code.
    async fn has_replicator_container_error(&self, prefix: &str) -> Result<bool, K8sError>;
}
