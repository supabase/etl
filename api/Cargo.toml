[package]
name = "api"
version = "0.1.0"
edition = "2024"

[lib]
path = "src/lib.rs"

[[bin]]
path = "src/main.rs"
name = "api"

[dependencies]
config = { workspace = true, features = ["utoipa"] }
postgres = { workspace = true, features = ["replication"] }
telemetry = { workspace = true }

actix-web = { workspace = true, features = ["macros", "http2"] }
actix-web-httpauth = { workspace = true }
anyhow = { workspace = true, features = ["std"] }
async-trait = { workspace = true }
aws-lc-rs = { workspace = true, features = ["alloc", "aws-lc-sys"] }
base64 = { workspace = true, features = ["std"] }
constant_time_eq = { workspace = true }
k8s-openapi = { workspace = true, features = ["latest"] }
kube = { workspace = true, features = [
    "runtime",
    "derive",
    "client",
    "rustls-tls",
] }
pg_escape = { workspace = true }
rand = { workspace = true, features = ["std"] }
reqwest = { workspace = true, features = ["json"] }
secrecy = { workspace = true }
sentry = { workspace = true, features = ["actix"] }
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true, features = ["std"] }
sqlx = { workspace = true, features = [
    "runtime-tokio-rustls",
    "macros",
    "postgres",
    "json",
    "migrate",
] }
thiserror = { workspace = true }
tokio = { workspace = true, features = ["rt-multi-thread", "macros"] }
tracing = { workspace = true, default-features = false }
tracing-actix-web = { workspace = true, features = ["emit_event_on_error"] }
utoipa = { workspace = true, features = ["actix_extras"] }
utoipa-swagger-ui = { workspace = true, features = ["actix-web", "reqwest"] }
uuid = { workspace = true, features = ["v4"] }

[dev-dependencies]
postgres = { workspace = true, features = ["test-utils", "sqlx"] }

insta = { workspace = true, features = ["json", "redactions"] }
