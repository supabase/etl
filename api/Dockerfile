# Build stage with cargo-chef for better layer caching
FROM lukemathwalker/cargo-chef:latest-rust-1.85.0-slim-bookworm AS chef
WORKDIR /app

# Install system dependencies
FROM chef AS planner
COPY . .
RUN cargo chef prepare --recipe-path recipe.json

# Build dependencies
FROM chef AS builder
RUN apt-get update && apt-get install -y pkg-config libssl-dev && rm -rf /var/lib/apt/lists/*
COPY --from=planner /app/recipe.json recipe.json
RUN cargo chef cook --release --recipe-path recipe.json

# Build application
COPY . .
ENV SQLX_OFFLINE=true
RUN cargo build --release -p api && \
    strip target/release/api

# Runtime stage with distroless for security
FROM gcr.io/distroless/cc-debian12:nonroot
WORKDIR /app

# Create non-root user (distroless already has nonroot user)
USER nonroot:nonroot

# Copy binary and configuration
COPY --from=builder /app/target/release/api ./api
COPY --chown=nonroot:nonroot api/configuration ./configuration

# Use exec form for proper signal handling
ENTRYPOINT ["./api"]
