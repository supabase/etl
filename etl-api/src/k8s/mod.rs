//! Kubernetes integration for the ETL API.
//!
//! This module contains the abstractions and implementations used by the HTTP
//! API to manage Kubernetes resources required by replicators (secrets, config
//! maps, stateful sets, and pods). Consumers should depend on the trait
//! [`K8sClient`] and avoid relying on a specific transport.
//!
//! The default client, [`http::HttpK8sClient`], is backed by the [`kube`]
//! crate and talks to the cluster using the ambient configuration (in-cluster
//! or local `~/.kube/config`). Keeping the abstraction in [`base`] lets us
//! swap implementations in tests and non-Kubernetes environments.
//!
//! See [`base`] for errors, pod phase mapping, and the client trait.

mod base;
pub mod http;

pub use base::*;
