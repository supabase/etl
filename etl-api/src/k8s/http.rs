use crate::k8s::{K8sClient, K8sError, PodPhase};
use async_trait::async_trait;
use base64::{Engine, prelude::BASE64_STANDARD};
use chrono::Utc;
use etl_config::Environment;
use k8s_openapi::api::{
    apps::v1::StatefulSet,
    core::v1::{ConfigMap, Pod, Secret},
};
use kube::{
    Client,
    api::{Api, DeleteParams, Patch, PatchParams},
};
use serde_json::json;
use tracing::debug;

/// Secret name suffix for the BigQuery service account key.
const BQ_SECRET_NAME_SUFFIX: &str = "bq-service-account-key";
/// Secret name suffix for the Postgres password.
const POSTGRES_SECRET_NAME_SUFFIX: &str = "postgres-password";
/// ConfigMap name suffix for the replicator configuration files.
const REPLICATOR_CONFIG_MAP_NAME_SUFFIX: &str = "replicator-config";
/// StatefulSet name suffix for the replicator workload.
const REPLICATOR_STATEFUL_SET_SUFFIX: &str = "replicator-stateful-set";
/// Application label suffix used to group resources.
const REPLICATOR_APP_SUFFIX: &str = "replicator-app";
/// Container name suffix for the replicator container.
const REPLICATOR_CONTAINER_NAME_SUFFIX: &str = "replicator";
/// Container name suffix for the Vector sidecar.
const VECTOR_CONTAINER_NAME_SUFFIX: &str = "vector";
/// Namespace where data-plane resources are created.
const DATA_PLANE_NAMESPACE: &str = "etl-data-plane";
/// Secret storing the Logflare API key.
const LOGFLARE_SECRET_NAME: &str = "replicator-logflare-api-key";
/// Docker image used for the Vector sidecar.
const VECTOR_IMAGE_NAME: &str = "timberio/vector:0.46.1-distroless-libc";
/// ConfigMap name containing the Vector configuration.
const VECTOR_CONFIG_MAP_NAME: &str = "replicator-vector-config";
/// Volume name for the replicator config file.
const REPLICATOR_CONFIG_FILE_VOLUME_NAME: &str = "replicator-config-file";
/// Volume name for the Vector config file.
const VECTOR_CONFIG_FILE_VOLUME_NAME: &str = "vector-config-file";
/// Secret storing the Sentry DSN.
const SENTRY_DSN_SECRET_NAME: &str = "replicator-sentry-dsn";
/// EmptyDir volume name used to share logs.
const LOGS_VOLUME_NAME: &str = "logs";
/// ConfigMap name providing trusted root certificates.
pub const TRUSTED_ROOT_CERT_CONFIG_MAP_NAME: &str = "trusted-root-certs-config";
/// Key inside the trusted root certificates ConfigMap.
pub const TRUSTED_ROOT_CERT_KEY_NAME: &str = "trusted_root_certs";
/// Pod template annotation used to trigger rolling restarts.
const RESTARTED_AT_ANNOTATION_KEY: &str = "etl.supabase.com/restarted-at";
/// Label used to identify replicator pods.
const REPLICATOR_APP_LABEL: &str = "etl-replicator-app";

/// Replicator memory limit tuned for `c6in.4xlarge` instances.
const REPLICATOR_MAX_MEMORY_PROD: &str = "500Mi";
/// Replicator CPU limit tuned for `c6in.4xlarge` instances.
const REPLICATOR_MAX_CPU_PROD: &str = "100m";
/// Replicator memory limit tuned for `t3.small` instances.
const REPLICATOR_MAX_MEMORY_STAGING: &str = "100Mi";
/// Replicator CPU limit tuned for `t3.small` instances.
const REPLICATOR_MAX_CPU_STAGING: &str = "100m";

/// Resource limits for a replicator pod.
struct ReplicatorResourceConfig {
    max_memory: &'static str,
    max_cpu: &'static str,
}

impl ReplicatorResourceConfig {
    /// Loads the runtime limits for the current environment.
    fn load(environment: &Environment) -> Result<Self, K8sError> {
        let config = match environment {
            Environment::Prod => Self {
                max_memory: REPLICATOR_MAX_MEMORY_PROD,
                max_cpu: REPLICATOR_MAX_CPU_PROD,
            },
            _ => Self {
                max_memory: REPLICATOR_MAX_MEMORY_STAGING,
                max_cpu: REPLICATOR_MAX_CPU_STAGING,
            },
        };

        Ok(config)
    }
}

/// HTTP-based implementation of [`K8sClient`].
///
/// The client is namespaced to the data-plane namespace and uses server-side
/// apply to keep resources in sync.
#[derive(Debug)]
pub struct HttpK8sClient {
    secrets_api: Api<Secret>,
    config_maps_api: Api<ConfigMap>,
    stateful_sets_api: Api<StatefulSet>,
    pods_api: Api<Pod>,
}

impl HttpK8sClient {
    /// Creates a new [`HttpK8sClient`] using the ambient Kubernetes config.
    ///
    /// Prefers in-cluster configuration and falls back to the local kubeconfig
    /// when running outside the cluster.
    pub async fn new() -> Result<HttpK8sClient, K8sError> {
        let client = Client::try_default().await?;

        let secrets_api: Api<Secret> = Api::namespaced(client.clone(), DATA_PLANE_NAMESPACE);
        let config_maps_api: Api<ConfigMap> = Api::namespaced(client.clone(), DATA_PLANE_NAMESPACE);
        let stateful_sets_api: Api<StatefulSet> =
            Api::namespaced(client.clone(), DATA_PLANE_NAMESPACE);
        let pods_api: Api<Pod> = Api::namespaced(client, DATA_PLANE_NAMESPACE);

        Ok(HttpK8sClient {
            secrets_api,
            config_maps_api,
            stateful_sets_api,
            pods_api,
        })
    }

    /// Helper function to handle delete operations that should ignore 404 errors
    /// but propagate other errors.
    fn handle_delete_with_404_ignore<T>(
        delete_result: Result<T, kube::Error>,
    ) -> Result<(), K8sError> {
        match delete_result {
            Ok(_) => Ok(()),
            Err(kube::Error::Api(er)) if er.code == 404 => Ok(()),
            Err(e) => Err(e.into()),
        }
    }
}

#[async_trait]
impl K8sClient for HttpK8sClient {
    async fn create_or_update_postgres_secret(
        &self,
        prefix: &str,
        postgres_password: &str,
    ) -> Result<(), K8sError> {
        debug!("patching postgres secret");

        let encoded_postgres_password = BASE64_STANDARD.encode(postgres_password);
        let postgres_secret_name = create_postgres_secret_name(prefix);
        let postgres_secret_json =
            create_postgres_secret_json(&postgres_secret_name, &encoded_postgres_password);
        let secret: Secret = serde_json::from_value(postgres_secret_json)?;

        let pp = PatchParams::apply(&postgres_secret_name);
        self.secrets_api
            .patch(&postgres_secret_name, &pp, &Patch::Apply(secret))
            .await?;

        Ok(())
    }

    async fn create_or_update_bq_secret(
        &self,
        prefix: &str,
        bq_service_account_key: &str,
    ) -> Result<(), K8sError> {
        debug!("patching bq secret");

        let encoded_bq_service_account_key = BASE64_STANDARD.encode(bq_service_account_key);
        let bq_secret_name = create_bq_secret_name(prefix);
        let bq_secret_json = create_bq_service_account_key_secret_json(
            &bq_secret_name,
            &encoded_bq_service_account_key,
        );
        let secret: Secret = serde_json::from_value(bq_secret_json)?;

        let pp = PatchParams::apply(&bq_secret_name);
        self.secrets_api
            .patch(&bq_secret_name, &pp, &Patch::Apply(secret))
            .await?;

        Ok(())
    }

    async fn delete_postgres_secret(&self, prefix: &str) -> Result<(), K8sError> {
        debug!("deleting postgres secret");

        let postgres_secret_name = create_postgres_secret_name(prefix);
        let dp = DeleteParams::default();
        Self::handle_delete_with_404_ignore(
            self.secrets_api.delete(&postgres_secret_name, &dp).await,
        )?;

        Ok(())
    }

    async fn delete_bq_secret(&self, prefix: &str) -> Result<(), K8sError> {
        debug!("deleting bq secret");

        let bq_secret_name = create_bq_secret_name(prefix);
        let dp = DeleteParams::default();
        Self::handle_delete_with_404_ignore(self.secrets_api.delete(&bq_secret_name, &dp).await)?;

        Ok(())
    }

    async fn get_config_map(&self, config_map_name: &str) -> Result<ConfigMap, K8sError> {
        debug!("getting config map");

        let config_map = match self.config_maps_api.get(config_map_name).await {
            Ok(config_map) => config_map,
            Err(e) => {
                return Err(e.into());
            }
        };

        Ok(config_map)
    }

    async fn create_or_update_config_map(
        &self,
        prefix: &str,
        base_config: &str,
        env_config: &str,
        environment: Environment,
    ) -> Result<(), K8sError> {
        debug!("patching config map");

        let env_config_file = format!("{environment}.yaml");
        let replicator_config_map_name = create_replicator_config_map_name(prefix);
        let config_map_json = create_replicator_config_map_json(
            &replicator_config_map_name,
            base_config,
            &env_config_file,
            env_config,
        );
        let config_map: ConfigMap = serde_json::from_value(config_map_json)?;

        let pp = PatchParams::apply(&replicator_config_map_name);
        self.config_maps_api
            .patch(&replicator_config_map_name, &pp, &Patch::Apply(config_map))
            .await?;

        Ok(())
    }

    async fn delete_config_map(&self, prefix: &str) -> Result<(), K8sError> {
        debug!("deleting config map");

        let replicator_config_map_name = create_replicator_config_map_name(prefix);
        let dp = DeleteParams::default();
        Self::handle_delete_with_404_ignore(
            self.config_maps_api
                .delete(&replicator_config_map_name, &dp)
                .await,
        )?;

        Ok(())
    }

    async fn create_or_update_stateful_set(
        &self,
        prefix: &str,
        replicator_image: &str,
        environment: Environment,
    ) -> Result<(), K8sError> {
        debug!("patching stateful set");

        let config = ReplicatorResourceConfig::load(&environment)?;

        let stateful_set_name = create_stateful_set_name(prefix);
        let replicator_app_name = create_replicator_app_name(prefix);
        let restarted_at_annotation = get_restarted_at_annotation_value();
        let replicator_container_name = create_replicator_container_name(prefix);
        let vector_container_name = create_vector_container_name(prefix);
        let postgres_secret_name = create_postgres_secret_name(prefix);
        let bq_secret_name = create_bq_secret_name(prefix);
        let replicator_config_map_name = create_replicator_config_map_name(prefix);

        let stateful_set_json = create_replicator_stateful_set_json(
            &stateful_set_name,
            &replicator_app_name,
            &restarted_at_annotation,
            &replicator_config_map_name,
            &vector_container_name,
            &config,
            &replicator_container_name,
            replicator_image,
            &environment,
            &postgres_secret_name,
            &bq_secret_name,
        );

        let stateful_set: StatefulSet = serde_json::from_value(stateful_set_json)?;

        let pp = PatchParams::apply(&stateful_set_name);
        self.stateful_sets_api
            .patch(&stateful_set_name, &pp, &Patch::Apply(stateful_set))
            .await?;

        Ok(())
    }

    async fn delete_stateful_set(&self, prefix: &str) -> Result<(), K8sError> {
        debug!("deleting stateful set");

        let stateful_set_name = create_stateful_set_name(prefix);
        let dp = DeleteParams::default();
        Self::handle_delete_with_404_ignore(
            self.stateful_sets_api.delete(&stateful_set_name, &dp).await,
        )?;

        Ok(())
    }

    async fn get_pod_phase(&self, prefix: &str) -> Result<PodPhase, K8sError> {
        debug!("getting pod status");

        let pod_name = create_pod_name(prefix);
        let pod = match self.pods_api.get(&pod_name).await {
            Ok(pod) => pod,
            Err(kube::Error::Api(er)) if er.code == 404 => return Ok(PodPhase::Succeeded),
            Err(e) => return Err(e.into()),
        };

        let phase = pod
            .status
            .map(|status| {
                let phase: PodPhase = status
                    .phase
                    .map(|phase| {
                        let phase: PodPhase = phase.as_str().into();
                        phase
                    })
                    .unwrap_or(PodPhase::Unknown);
                phase
            })
            .unwrap_or(PodPhase::Unknown);

        Ok(phase)
    }

    async fn has_replicator_container_error(&self, prefix: &str) -> Result<bool, K8sError> {
        debug!("checking for replicator container error");

        let pod_name = create_pod_name(prefix);
        let pod = match self.pods_api.get(&pod_name).await {
            Ok(pod) => pod,
            Err(kube::Error::Api(er)) if er.code == 404 => return Ok(false),
            Err(e) => return Err(e.into()),
        };

        let replicator_container_name = create_replicator_container_name(prefix);

        // Find the replicator container status
        let container_status = pod.status.and_then(|status| {
            status.container_statuses.and_then(|container_statuses| {
                container_statuses
                    .iter()
                    .find(|cs| cs.name == replicator_container_name)
                    .cloned()
            })
        });

        let Some(container_status) = container_status else {
            return Ok(false);
        };

        // Check last terminated state for non-zero exit code
        if let Some(last_state) = &container_status.last_state
            && let Some(terminated) = &last_state.terminated
        {
            return Ok(terminated.exit_code != 0);
        }

        Ok(false)
    }
}

fn create_postgres_secret_name(prefix: &str) -> String {
    format!("{prefix}-{POSTGRES_SECRET_NAME_SUFFIX}")
}

fn create_bq_secret_name(prefix: &str) -> String {
    format!("{prefix}-{BQ_SECRET_NAME_SUFFIX}")
}

fn create_replicator_config_map_name(prefix: &str) -> String {
    format!("{prefix}-{REPLICATOR_CONFIG_MAP_NAME_SUFFIX}")
}

fn create_stateful_set_name(prefix: &str) -> String {
    format!("{prefix}-{REPLICATOR_STATEFUL_SET_SUFFIX}")
}

fn create_pod_name(prefix: &str) -> String {
    format!("{prefix}-{REPLICATOR_STATEFUL_SET_SUFFIX}-0")
}

fn create_replicator_app_name(prefix: &str) -> String {
    format!("{prefix}-{REPLICATOR_APP_SUFFIX}")
}

fn create_replicator_container_name(prefix: &str) -> String {
    format!("{prefix}-{REPLICATOR_CONTAINER_NAME_SUFFIX}")
}

fn create_vector_container_name(prefix: &str) -> String {
    format!("{prefix}-{VECTOR_CONTAINER_NAME_SUFFIX}")
}

fn create_postgres_secret_json(
    secret_name: &str,
    encoded_postgres_password: &str,
) -> serde_json::Value {
    json!({
      "apiVersion": "v1",
      "kind": "Secret",
      "metadata": {
        "name": secret_name,
        "namespace": DATA_PLANE_NAMESPACE,
      },
      "type": "Opaque",
      "data": {
        "password": encoded_postgres_password,
      }
    })
}

fn create_bq_service_account_key_secret_json(
    secret_name: &str,
    encoded_bq_service_account_key: &str,
) -> serde_json::Value {
    json!({
      "apiVersion": "v1",
      "kind": "Secret",
      "metadata": {
        "name": secret_name,
        "namespace": DATA_PLANE_NAMESPACE,
      },
      "type": "Opaque",
      "data": {
        "service-account-key": encoded_bq_service_account_key,
      }
    })
}

fn create_replicator_config_map_json(
    config_map_name: &str,
    base_config: &str,
    env_config_file: &str,
    env_config: &str,
) -> serde_json::Value {
    json!({
      "kind": "ConfigMap",
      "apiVersion": "v1",
      "metadata": {
        "name": config_map_name,
        "namespace": DATA_PLANE_NAMESPACE,
      },
      "data": {
        "base.yaml": base_config,
        env_config_file: env_config,
      }
    })
}

#[expect(clippy::too_many_arguments)]
fn create_replicator_stateful_set_json(
    stateful_set_name: &str,
    replicator_app_name: &str,
    restarted_at_annotation: &str,
    replicator_config_map_name: &str,
    vector_container_name: &str,
    config: &ReplicatorResourceConfig,
    replicator_container_name: &str,
    replicator_image: &str,
    environment: &Environment,
    postgres_secret_name: &str,
    bq_secret_name: &str,
) -> serde_json::Value {
    json!({
      "apiVersion": "apps/v1",
      "kind": "StatefulSet",
      "metadata": {
        "name": stateful_set_name,
        "namespace": DATA_PLANE_NAMESPACE,
      },
      "spec": {
        "replicas": 1,
        "selector": {
          "matchLabels": {
            "app-name": replicator_app_name,
          }
        },
        "template": {
          "metadata": {
            "labels": {
              "app-name": replicator_app_name,
              "app": REPLICATOR_APP_LABEL
            },
            "annotations": {
              // Attach template annotations (e.g., restart checksum) to trigger a rolling restart
              RESTARTED_AT_ANNOTATION_KEY: restarted_at_annotation,
            }
          },
          "spec": {
            "volumes": [
              {
                "name": REPLICATOR_CONFIG_FILE_VOLUME_NAME,
                "configMap": {
                  "name": replicator_config_map_name
                }
              },
              {
                "name": VECTOR_CONFIG_FILE_VOLUME_NAME,
                "configMap": {
                  "name": VECTOR_CONFIG_MAP_NAME
                }
              },
              {
                "name": LOGS_VOLUME_NAME,
                "emptyDir": {}
              }
            ],
            // Allow scheduling onto nodes tainted with `nodeType=workloads`.
            "tolerations": [
              {
                "key": "nodeType",
                "operator": "Equal",
                "value": "workloads",
                "effect": "NoSchedule"
              }
            ],
            // Pin pods to nodes labeled with `nodeType=workloads`.
            "nodeSelector": {
              "nodeType": "workloads"
            },
            // We want to wait at most 5 minutes before K8S sends a `SIGKILL` to the containers,
            // this way we let the system finish any in-flight transaction, if there are any.
            "terminationGracePeriodSeconds": 300,
            "initContainers": [
              {
                "name": vector_container_name,
                "image": VECTOR_IMAGE_NAME,
                "restartPolicy": "Always",
                "env": [
                  {
                    "name": "LOGFLARE_API_KEY",
                    "valueFrom": {
                      "secretKeyRef": {
                        "name": LOGFLARE_SECRET_NAME,
                        "key": "key"
                      }
                    }
                  }
                ],
                "resources": {
                  "limits": {
                    "memory": config.max_memory,
                    "cpu": config.max_cpu,
                  },
                  "requests": {
                    "memory": config.max_memory,
                    "cpu": config.max_cpu,
                  }
                },
                "volumeMounts": [
                  {
                    "name": VECTOR_CONFIG_FILE_VOLUME_NAME,
                    "mountPath": "/etc/vector"
                  },
                  {
                    "name": LOGS_VOLUME_NAME,
                    "mountPath": "/var/log"
                  }
                ]
              }
            ],
            "containers": [
              {
                "name": replicator_container_name,
                "image": replicator_image,
                "ports": [
                  {
                    "name": "metrics",
                    "containerPort": 9000,
                    "protocol": "TCP"
                  }
                ],
                "env": [
                  {
                    "name": "APP_ENVIRONMENT",
                    "value": environment.to_string()
                  },
                  {
                      "name": "APP_VERSION",
                      //TODO: set APP_VERSION to proper version instead of the replicator image name
                      "value": replicator_image
                  },
                  {
                    "name": "APP_SENTRY__DSN",
                    "valueFrom": {
                      "secretKeyRef": {
                        "name": SENTRY_DSN_SECRET_NAME,
                        "key": "dsn"
                      }
                    }
                  },
                  {
                    "name": "APP_PIPELINE__PG_CONNECTION__PASSWORD",
                    "valueFrom": {
                      "secretKeyRef": {
                        "name": postgres_secret_name,
                        "key": "password"
                      }
                    }
                  },
                  {
                    "name": "APP_DESTINATION__BIG_QUERY__SERVICE_ACCOUNT_KEY",
                    "valueFrom": {
                      "secretKeyRef": {
                        "name": bq_secret_name,
                        "key": "service-account-key"
                      }
                    }
                  }
                ],
                "volumeMounts": [
                  {
                    "name": REPLICATOR_CONFIG_FILE_VOLUME_NAME,
                    "mountPath": "/app/configuration"
                  },
                  {
                    "name": LOGS_VOLUME_NAME,
                    "mountPath": "/app/logs"
                  },
                ]
              }
            ]
          }
        }
      }
    })
}

fn get_restarted_at_annotation_value() -> String {
    let now = Utc::now();
    // We use nanoseconds to decrease the likelihood of generating the same annotation in sequence,
    // which would not result in a restart.
    now.to_rfc3339_opts(chrono::SecondsFormat::Nanos, true)
}

#[cfg(test)]
mod tests {
    use super::*;
    use etl_config::{
        SerializableSecretString,
        shared::{
            BatchConfig, DestinationConfig, PgConnectionConfig, PipelineConfig, ReplicatorConfig,
            TlsConfig,
        },
    };
    use insta::assert_json_snapshot;

    const TENANT_ID: &str = "abcdefghijklmnopqrst";

    fn create_k8s_object_prefix(tenant_id: &str, replicator_id: i64) -> String {
        format!("{tenant_id}-{replicator_id}")
    }

    #[test]
    fn test_create_postgres_secret_json() {
        let secret_name = "test-secret";
        let encoded_postgres_password = "dGVzdC1wYXNzd29yZA==";

        let secret_json = create_postgres_secret_json(secret_name, encoded_postgres_password);

        assert_json_snapshot!(secret_json);

        let _secret: Secret = serde_json::from_value(secret_json).unwrap();
    }

    #[test]
    fn test_create_bq_service_account_key_secret_json() {
        let secret_name = "test-secret";
        let encoded_bq_service_account_key = "ewogICJrZXkiOiAidmFsdWUiCn0=";

        let secret_json =
            create_bq_service_account_key_secret_json(secret_name, encoded_bq_service_account_key);

        assert_json_snapshot!(secret_json);

        let _secret: Secret = serde_json::from_value(secret_json).unwrap();
    }

    #[test]
    fn test_create_replicator_config_map_json() {
        let prefix = create_k8s_object_prefix(TENANT_ID, 42);
        let replicator_config_map_name = create_replicator_config_map_name(&prefix);
        let environment = Environment::Prod;
        let env_config_file = format!("{environment}.yaml");
        let base_config = "";
        let replicator_config = ReplicatorConfig {
            destination: DestinationConfig::BigQuery {
                project_id: "project-id".to_string(),
                dataset_id: "dataset-id".to_string(),
                service_account_key: SerializableSecretString::from("sa-key".to_string()),
                max_staleness_mins: None,
                max_concurrent_streams: 4,
            },
            pipeline: PipelineConfig {
                id: 42,
                publication_name: "all-pub".to_string(),
                pg_connection: PgConnectionConfig {
                    host: "localhost".to_string(),
                    port: 5432,
                    name: "postgres".to_string(),
                    username: "postgres".to_string(),
                    password: None,
                    tls: TlsConfig {
                        trusted_root_certs: "".to_string(),
                        enabled: false,
                    },
                },
                batch: BatchConfig {
                    max_size: 10_000,
                    max_fill_ms: 1_000,
                },
                table_error_retry_delay_ms: 500,
                table_error_retry_max_attempts: 3,
                max_table_sync_workers: 4,
            },
            sentry: None,
            supabase: None,
        };
        let env_config = serde_json::to_string(&replicator_config).unwrap();

        let config_map_json = create_replicator_config_map_json(
            &replicator_config_map_name,
            base_config,
            &env_config_file,
            &env_config,
        );

        assert_json_snapshot!(config_map_json);

        let _config_map: ConfigMap = serde_json::from_value(config_map_json).unwrap();
    }

    #[test]
    fn test_create_replicator_stateful_set_json() {
        let prefix = create_k8s_object_prefix(TENANT_ID, 42);
        let stateful_set_name = create_stateful_set_name(&prefix);
        let replicator_app_name = create_replicator_app_name(&prefix);
        let restarted_at_annotation = "2025-10-09T16:02:24.127400000Z";
        let replicator_container_name = create_replicator_container_name(&prefix);
        let vector_container_name = create_vector_container_name(&prefix);
        let postgres_secret_name = create_postgres_secret_name(&prefix);
        let bq_secret_name = create_bq_secret_name(&prefix);
        let replicator_config_map_name = create_replicator_config_map_name(&prefix);
        let environment = Environment::Prod;
        let config = ReplicatorResourceConfig::load(&environment).unwrap();
        let replicator_image = "ramsup/etl-replicator:2a41356af735f891de37d71c0e1a62864fe4630e";

        let stateful_set_json = create_replicator_stateful_set_json(
            &stateful_set_name,
            &replicator_app_name,
            restarted_at_annotation,
            &replicator_config_map_name,
            &vector_container_name,
            &config,
            &replicator_container_name,
            replicator_image,
            &environment,
            &postgres_secret_name,
            &bq_secret_name,
        );

        assert_json_snapshot!(stateful_set_json);

        let _stateful_set: StatefulSet = serde_json::from_value(stateful_set_json).unwrap();
    }
}
