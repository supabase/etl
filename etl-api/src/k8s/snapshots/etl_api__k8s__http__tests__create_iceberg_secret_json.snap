---
source: etl-api/src/k8s/http.rs
expression: secret_json
---
{
  "apiVersion": "v1",
  "data": {
    "catalog-token": "ZXlKMGVYQWlPaUpLVjFRaUxDSmhiR2NpT2lKRlV6STFOaUlzSW10cFpDSTZJakZrTnpGak1HRXlObUl4TURGak9EUTVaVGt4Wm1RMU5qZGpZakE1TlRKbUluMC5leUpsZUhBaU9qSXdOekEzTVRjeE5qQXNJbWxoZENJNk1UYzFOakUwTlRFMU1Dd2lhWE56SWpvaWMzVndZV0poYzJVaUxDSnlaV1lpT2lKaFltTmtaV1puYUdscWJHdHRibTl3Y1hKemRDSXNJbkp2YkdVaU9pSnpaWEoyYVdObFgzSnZiR1VpZlEuWWRUV2trSXZ3alNrWG90M05DMDd4eWpQakdXUU1OekxxNUVQenVtenJkTHp1SHJqLXp1ekktbmx5UXRRNVY3Z1phdXlzbS13R3dtcHp0UlhmUGMzQVE=",
    "s3-access-key-id": "Y2FlNGY0NjliNTY5MjJhMTNmMzNiNjM3YTNjMWU2ZjI=",
    "s3-secret-access-key": "NDUyOWE3ZmMwNzY2NDBjODRiZTgzZGJiNGMyNDI3MTNhOTk0MzE5OTBjYzJmMzIzMGM4MzVjOGJmZjAzYWE2ZQ=="
  },
  "kind": "Secret",
  "metadata": {
    "name": "abcdefghijklmnopqrst-42-iceberg",
    "namespace": "etl-data-plane"
  },
  "type": "Opaque"
}
